### ISSUE TITLE
BUCK doesn't support kapt's generated option?

### ISSUE THREAD

**Issue by hzsweers** [facebook/buck#1891]
Trying to use Moshi's new Kotlin code gen feature, it seems that while the processor *runs*, the code it generates is going into a black hole somewhere.

It uses kapt's `generated` option to know where to generate code. In gradle, this ends up roughly in `$buildDir/generated/source/kaptKotlin/main`. For kotlin, this seems to go nowhere. Not sure if the option just isn't being set or what.

See: https://kotlinlang.org/docs/reference/kapt.html#generating-kotlin-sources

**Comment by hzsweers** [comment]
CC @tyvsmith @kageiit

**Comment by hzsweers** [comment]
I do see it wired in

https://github.com/facebook/buck/blob/06ecade50006cfb4b0a5e63a04eea0d63792b91c/src/com/facebook/buck/jvm/kotlin/KotlincToJarStepFactory.java#L88

https://github.com/facebook/buck/blob/06ecade50006cfb4b0a5e63a04eea0d63792b91c/src/com/facebook/buck/jvm/kotlin/KotlincToJarStepFactory.java#L137-L138

And those directories are getting generated, but there's no output in them

**Comment by hzsweers** [comment]
After some grep magic, I'm seeing them, but in really random places

If I have a project under `/libraries/foundation/modelgen/thrift-utilities`

they end up duplicated in several places

```
./buck-out/gen/libraries/foundation/modelgen/thrifty-utilities/lib__src_main____working_directory/com/uber/thrifty/ThriftFieldJsonAdapter.kt

./buck-out/annotation/libraries/foundation/modelgen/thrifty-utilities/__src_main_kapt_generated__/com/uber/thrifty/ThriftFieldJsonAdapter.kt

./buck-out/bin/libraries/foundation/modelgen/thrifty-utilities/__src_main_gen_sources__/com/uber/thrifty/ThriftFieldJsonAdapter.kt

./build/buildScript/project-build/libraries/foundation/modelgen/thrifty-utilities/build/generated/source/kaptKotlin/main/com/uber/thrifty/ThriftFieldJsonAdapter.kt

./build/buildScript2/project-build/libraries/foundation/modelgen/thrifty-utilities/build/generated/source/kaptKotlin/main/com/uber/thrifty/ThriftFieldJsonAdapter.kt
```

Does this imply the processor is being invoked multiple times with different output locations?

**Comment by hzsweers** [comment]
One of them does appear on the classpath, but not sure how to see which

**Comment by hzsweers** [comment]
CC @thalescm

**Comment by thalescm** [comment]
Hey @hzsweers ,
Kapt can pottentially generate stuff in three different directories:
`kapt.kotlin.generated` option path, which kotlin APs should point out to.
`sources`, which is normally where java APs output appears
`classes`, if you're generating class or resources files
Check more out [here](https://kotlinlang.org/docs/reference/kapt.html#using-in-cli).

What happens on buck is that kapt has [four](https://github.com/facebook/buck/issues/956#issuecomment-309080611) steps for running (decribed in the link). But you can not know for sure, while on buck, if it will, or will not generate anything.
* So we [run kapt](https://github.com/facebook/buck/blob/master/src/com/facebook/buck/jvm/kotlin/KotlincToJarStepFactory.java#L178), which is the file you see at `./buck-out/annotation/libraries/foundation/modelgen/thrifty-utilities/__src_main_kapt_generated__`
* [copy the output](https://github.com/facebook/buck/blob/master/src/com/facebook/buck/jvm/kotlin/KotlincToJarStepFactory.java#L196-L204) from those directories to a tmp directory, which is the file you see at `./buck-out/bin/libraries/foundation/modelgen/thrifty-utilities/__src_main_gen_sources__`
* [Zip that file](https://github.com/facebook/buck/blob/master/src/com/facebook/buck/jvm/kotlin/KotlincToJarStepFactory.java#L209) and add it [to the sources](https://github.com/facebook/buck/blob/master/src/com/facebook/buck/jvm/kotlin/KotlincToJarStepFactory.java#L237). Kotlinc nor javac accepts zipfiles as entry, so we have to unzip them, and we do that on the [project working directory](), which is the file you see at `./buck-out/gen/libraries/foundation/modelgen/thrifty-utilities/lib__src_main____working_directory`


We should probably delete both files at `annotation` and `bin` directories at the and of the step, would be cleaner. To make IDE see the files generated by buck kapt, you could add a config for gradle that src files contain `buck-out/path/to/your/project/gen/kapt`.
The other two paths where they appear are gradle outputs.

Hopefully that cleared how the flow works, but anything else please ask.
Answering the final question, that doesn't mean the processor is being called multiple times. 
But it does indicate you're running kapt both with buck and with gradle.

**Comment by hzsweers** [comment]
Agreed on deleting being the best path forward to clear up confusion. Will try the gradle config (I'm a bit worried we'd have to end up pointing somewhere in buck-out), but shouldn't buck project pick this up correctly in native buck?

**Comment by thalescm** [comment]
You mean a buck project in the AS? I don't think so, don't know and I'm not that familiar to how buck will generate the project files. But when implementing kapt I didn't touched on that part, so probably the IDE won't recognize those files for autocomplete and etc...
Probably a good idea the create an issue specific for that, if that's the case

**Comment by hzsweers** [comment]
Hey - it would be great if someone from the buck team could acknowledge this. It's a major blocker for kotlin-first annotation processors
